crates/mpz-garble/src/protocol/semihonest/evaluator.rs

```diff
@@ -0,0 +1,358 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use hashbrown::HashMap;
+use tokio::sync::Mutex;
+use utils::{
+    filter_drain::FilterDrain,
+    range::{Disjoint, RangeSet},
+};
+
+use mpz_common::{
+    scoped_futures::{ScopedBoxFuture, ScopedFutureExt},
+    Context, Flush,
+};
+use mpz_core::{bitvec::BitVec, Block};
+use mpz_garble_core::{evaluate_garbled_circuits, EvaluatorOutput, GarbledCircuit};
+use mpz_memory_core::{binary::Binary, DecodeFuture, Memory, Slice, View};
+use mpz_ot::cot::COTReceiver;
+use mpz_vm_core::{Call, Execute, Vm};
+
+use crate::{
+    evaluator::receive_garbled_circuit,
+    store::{EvaluatorStore, EvaluatorStoreError},
+};
+
+type Result<T, E = EvaluatorError> = core::result::Result<T, E>;
+type Error = EvaluatorError;
+
+/// Semi-honest evaluator.
+#[derive(Debug)]
+pub struct Evaluator<COT> {
+    store: Arc<Mutex<EvaluatorStore<COT>>>,
+    call_stack: Vec<(Call, Slice)>,
+    preprocessed: HashMap<Slice, (Call, GarbledCircuit)>,
+}
+
+impl<COT> Evaluator<COT> {
+    /// Creates a new evaluator.
+    pub fn new(cot: COT) -> Self {
+        Self {
+            store: Arc::new(Mutex::new(EvaluatorStore::new(cot))),
+            call_stack: Vec::new(),
+            preprocessed: HashMap::new(),
+        }
+    }
+
+    fn take_preprocess_calls(&mut self) -> Vec<(Call, Slice)> {
+        let mut idx_outputs = RangeSet::default();
+        self.call_stack
+            // Extract calls which have no dependencies on other prior calls.
+            .filter_drain(|(call, output)| {
+                if call
+                    .inputs()
+                    .iter()
+                    .all(|input| input.to_range().is_disjoint(&idx_outputs))
+                {
+                    idx_outputs |= output.to_range();
+                    true
+                } else {
+                    false
+                }
+            })
```

Comment:
Isn't that dependent on the order of calls in `call_stack`. So don't you have to build `idx_outputs` in a first iteration and then `filter_drain` in a second one?

Reply:
You want to collect calls which have no dependencies on other calls. To do this you check if all inputs of a call are disjoint with `idx_outputs`. But in the same loop `idx_outputs` is built. 

For example you have two calls. Let's name them 1 and 2 and assume call 2 depends on call 1. Now what happens if call 2 is **before** call 1 in `call_stack`. Then the closure will return **true** for call 2, although it depends on call 1. This is because `idx_outputs` does not yet contain the outputs of call 1.